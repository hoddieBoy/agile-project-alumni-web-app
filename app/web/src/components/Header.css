.header {
    display: flex;
    justify-content: space-between;
    align-items: center;
    background-color: #002855;
    padding: 10px;
}

.header-left,
.header-right {
    display: flex;
    align-items: center;
}

.header .navbar-brand {
    display: flex;
    align-items: center;
    margin-right: 20px;
}

.header .navbar-brand img {
    max-height: 80px;
    object-fit: contain;
    margin: -15px 0;
}

.header .navbar-search {
    padding: 5px 10px;
    border-radius: 4px;
    border: none;
}

.header .nav-item {
    color: white;
    text-decoration: none;
    padding: 8px 12px;
    border-radius: 4px;
    transition: background-color 0.3s;
}

.header .nav-item:hover {
    background-color: rgba(255, 255, 255, 0.1);
}

.header .btn {
    background-color: #ffffff;
    color: #002855;
    border: none;
    padding: 8px 16px;
    border-radius: 4px;
    cursor: pointer;
    transition: background-color 0.3s, color 0.3s;
}

.header .btn:hover {
    background-color: #f0f0f0;
    color: #002855;
}

@media (max-width: 768px) {
    .header {
        flex-direction: column;
        align-items: flex-start;
    }

    .header-left,
    .header-right {
        width: 100%;
        justify-content: space-between;
    }

    .header .navbar-search {
        width: 100%;
        margin-bottom: 10px;
    }

    .header .nav-item,
    .header .btn {
        width: 100%;
        text-align: center;
        margin-bottom: 5px;
    }
}