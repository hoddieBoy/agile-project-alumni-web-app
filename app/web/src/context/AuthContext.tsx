import React, {createContext, ReactNode, useEffect, useState} from 'react';

// Define the AuthContextType interface
export interface AuthContextType {
    isAuthenticated: boolean;
    login: () => void;
    logout: () => void;
}

// Create the AuthContext with default values
const AuthContext = createContext<AuthContextType>({
    isAuthenticated: false,
    login: () => {
    },
    logout: () => {
    }
});

const AuthProvider: React.FC<{ children: ReactNode }> = ({children}) => {
    const [isAuthenticated, setIsAuthenticated] = useState<boolean>(() => {
        const savedAuthState = localStorage.getItem('isAuthenticated');
        return savedAuthState ? JSON.parse(savedAuthState) : false;
    });

    useEffect(() => {
        localStorage.setItem('isAuthenticated', JSON.stringify(isAuthenticated));
    }, [isAuthenticated]);

    const login = () => setIsAuthenticated(true);
    const logout = () => setIsAuthenticated(false);

    return (
        <AuthContext.Provider value={{isAuthenticated, login, logout}}>
            {children}
        </AuthContext.Provider>
    );
};

export {AuthContext, AuthProvider};