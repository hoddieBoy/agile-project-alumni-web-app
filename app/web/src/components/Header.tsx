import React from "react";
import {deleteCookie} from "utils/Cookie";
import {redirect} from "react-router-dom";
import "components/Header.css";

export default function Header() {
    const handleLogout = () => {
        deleteCookie('token');
        redirect('/login');
    }

    return (
        <header className="header">
            <div className="header-left">
                <a href="#" className="navbar-brand">
                    <img
                        src="https://www.imt-atlantique.fr/sites/default/files/Images/Ecole/charte-graphique/IMT_Atlantique_logo_RVB_Negatif_Baseline_400x272.png"
                        alt="IMT Atlantique Logo"
                    />
                </a>
                <input type="text" className="form-control navbar-search" placeholder="Search"/>
            </div>
            <nav className="header-right">
                <a href="#" className="nav-link">Search</a>
                <a href="#" className="nav-link">Import/Export</a>
                <a href="#" className="nav-link">Gestion Utilisateurs</a>
                <button className="btn" onClick={handleLogout}>Log out</button>
            </nav>
        </header>
    );
}