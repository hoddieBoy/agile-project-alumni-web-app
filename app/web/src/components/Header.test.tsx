// Header.test.tsx
import React from 'react';
import {fireEvent, render, screen} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Header from './Header';
import {AuthContext, AuthContextType} from 'context/AuthContext';

// Create a custom render method that includes the AuthContext provider
const renderWithAuthContext = (contextValue: Partial<AuthContextType>, component: React.ReactNode) => {
    return render(
        <AuthContext.Provider value={contextValue as AuthContextType}>
            {component}
        </AuthContext.Provider>
    );
};

describe('Header component', () => {
    let mockLogout: jest.Mock;

    beforeEach(() => {
        mockLogout = jest.fn();
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('renders without crashing', () => {
        const {asFragment} = renderWithAuthContext({
            logout: mockLogout,
            isAuthenticated: true,
            user: {username: 'admin', roles: ['ROLE_ADMIN']}
        }, <Header/>);
        expect(asFragment()).toMatchSnapshot();
    });

    describe('renders the header elements', () => {
        it('renders the logo with correct attributes', () => {
            renderWithAuthContext({
                logout: mockLogout,
                isAuthenticated: true,
                user: {username: 'admin', roles: ['ROLE_ADMIN']}
            }, <Header/>);
            const logo = screen.getByAltText('IMT Atlantique Logo');
            expect(logo).toBeInTheDocument();
            expect(logo).toHaveAttribute('src', 'https://www.imt-atlantique.fr/sites/default/files/Images/Ecole/charte-graphique/IMT_Atlantique_logo_RVB_Negatif_Baseline_400x272.png');
        });

        it('renders the search input with correct attributes', () => {
            renderWithAuthContext({
                logout: mockLogout,
                isAuthenticated: true,
                user: {username: 'admin', roles: ['ROLE_ADMIN']}
            }, <Header/>);
            const searchInput = screen.getByPlaceholderText('Search');
            expect(searchInput).toBeInTheDocument();
            expect(searchInput).toHaveAttribute('type', 'text');
        });

        it('renders navigation links', () => {
            renderWithAuthContext({
                logout: mockLogout,
                isAuthenticated: true,
                user: {username: 'admin', roles: ['ROLE_ADMIN']}
            }, <Header/>);
            const links = ['Search', 'Import-Export', 'Stat', 'Gestion Utilisateurs'];
            links.forEach(link => {
                const navLink = screen.getByText(link);
                expect(navLink).toBeInTheDocument();
                expect(navLink).toHaveAttribute('href', `/${link.toLowerCase().replace(/\s/g, '-')}`);
            });
        });

        it('renders the logout button', () => {
            renderWithAuthContext({
                logout: mockLogout,
                isAuthenticated: true,
                user: {username: 'admin', roles: ['ROLE_ADMIN']}
            }, <Header/>);
            const logoutButton = screen.getByRole('button', {name: /log out/i});
            expect(logoutButton).toBeInTheDocument();
            expect(logoutButton).toHaveClass('btn btn-primary');
        });
    });

    it('calls logout function when logout button is clicked', () => {
        renderWithAuthContext({
            logout: mockLogout,
            isAuthenticated: true,
            user: {username: 'admin', roles: ['ROLE_ADMIN']}
        }, <Header/>);
        const logoutButton = screen.getByRole('button', {name: /log out/i});
        fireEvent.click(logoutButton);
        expect(mockLogout).toHaveBeenCalled();
    });

    it('does not render "Gestion Utilisateurs" link if user is not an admin', () => {
        renderWithAuthContext({
            logout: mockLogout,
            isAuthenticated: true,
            user: {username: 'user', roles: ['ROLE_USER']}
        }, <Header/>);
        const gestionLink = screen.queryByText('Gestion Utilisateurs');
        expect(gestionLink).not.toBeInTheDocument();
    });
});
