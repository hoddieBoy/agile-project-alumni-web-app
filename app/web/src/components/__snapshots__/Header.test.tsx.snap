// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component renders without crashing 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <div
      class="header-left"
    >
      <a
        aria-label="IMT Atlantique"
        class="navbar-brand"
        href="/"
      >
        <img
          alt="IMT Atlantique Logo"
          src="https://www.imt-atlantique.fr/sites/default/files/Images/Ecole/charte-graphique/IMT_Atlantique_logo_RVB_Negatif_Baseline_400x272.png"
        />
      </a>
      <input
        aria-label="Search"
        class="form-control navbar-search"
        placeholder="Search"
        type="text"
      />
    </div>
    <nav
      class="header-right gap-3"
    >
      <a
        class="nav-item nav-link"
        href="/search"
      >
        Search
      </a>
      <a
        class="nav-item nav-link"
        href="/import-export"
      >
        Import-Export
      </a>
      <a
        class="nav-item nav-link"
        href="/gestion-utilisateurs"
      >
        Gestion Utilisateurs
      </a>
      <a
        class="nav-item nav-link"
        href="/stat"
      >
        Stat
      </a>
      <button
        class="btn btn-primary"
        type="button"
      >
        Log out
      </button>
    </nav>
  </header>
</DocumentFragment>
`;
